designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,527 @@
+        rules: {
+            semi: "error"
+        }
+    }
+
+];
+```
+
+When an extended config and the base config both have multiple `files` entries, then the result is a `files` entry containing all combinations. For example:
+
+```js
+export default [
+    {
+        files: ["src/**", "lib/**"],
+        extends: [{ files: ["**/*.js", "**/*.mjs"] }]
+    }
+];
+```
+
+This config would be calculated as the following:
+
+```js
+export default [
+    {
+        files: [
+            ["src/**", "**/*.js"],
+            ["lib/**", "**/*.js"],
+            ["src/**", "**/*.mjs"],
+            ["lib/**", "**/*.mjs"]
+        ]
+    }
+];
+```
+
+Notes:
+
+1. The `files` and `ignores` values from the base config always come first in the calculated `files` and `ignores`. If there's not a match, then there's no point in continuing on to use the patterns from the extended configs.
+1. The `name` key is generated for calculated config objects so that it indicates the inheritance from another config. The extended configs don't exist in the final representation of the config array.
+1. These behaviors differ from the [typescript-eslint extends helper](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/typescript-eslint/src/config-helper.ts).
+
+#### Extending Arrays
+
+Arrays can also be used in `extends` (to eliminate the guesswork of what type a config is). When evaluating `extends`, ESLint internally calls `.flat(Infinity)` on the array and then processes the config objects as discussed in the previous example. Consider the following:
+
+```js
+import js from "@eslint/js";
+import example from "eslint-plugin-example";
+
+export default [
+
+    {
+        files: ["**/src/*.js"],
+        extends: [
+            [js.configs.recommended, example.configs.recommended]
+        ]
+    }
+
+];
+```
+
+This is equivalent to the following:
+
+```js
+import js from "@eslint/js";
+import example from "eslint-plugin-example";
+
+export default [
+
+    {
+        files: ["**/src/*.js"],
+        extends: [js.configs.recommended, example.configs.recommended]
+    }
+
+];
+```
+
+The extended objects are evaluated in the same order and result in the same final config.
+
+#### Extending Named Configs
+
+While the previous examples are an improvement over current situation, we can go a step further and restore the use of named configs by allowing strings inside of the `extends` array. When provided, the string must refer to a config contained in a plugin. For example:
+
+```js
+import js from "@eslint/js";
+
+export default [
+
+    {
+        plugins: {
+            js
+        },
+        files: ["**/src/*.js"],
+        extends: ["js/recommended"]
+    }
+
+];
+```
+
+Here, `js/recommended` refers to the plugin defined as `js`. Internally, ESLint will look up the plugin with the name `js`, looks at the `configs` key, and retrieve the `recommended` key as a replacement for the string `"js/recommended"`. 
```

Comment:
A vote against `extends` strings: I don't think they reasonably be modeled in a type system.

One of the points I really like about flat config is how it's much more statically analyzable than legacy/eslintrc. Other than `rules` entries, roughly everything can be seen and helped with by editor intellisense / type checking. It being _"just JavaScript"_ means that as users type in files like `eslint.config.*`, they get TypeScript assistance.

Having ~~"magic"~~ not-easily-type-checked string constants is hard to model in the type system. I think that's reflective of how they need to be reasoned about. Users need to now memorize some "arbitrary" (unique to ESLint) system for mapping strings to plugins -> configs. As in eslintrc, those arbitrary systems are an extra burden to learn -- and extra burdensome because of the lessened editor help.

Reply:
I understand your point, I'm just not sure that modeling in the type system is as necessary as it seems. We went ten years with eslintrc and people seemed to do just fine. Worse case there's an error if you access a string that doesn't exist.

Reply:
> and people seemed to do just fine

**_STRONGLY_** disagree.

People fumbled around and "made it work". They didn't do "just fine"!
Many people grumbled when the switch to `.js` eslintrcs was enforced because it meant that they no longer had a JSONSchema to give them IDE features on their configs.

People would often write their config file like this so that they would get a type-checked config with IDE features:
```ts
/** @type {import('eslint').Linter.Config} */
module.exports = { ... };
```
[Here's an example query showing that there's still ~3000 eslintrcs across github that exist and do this](https://sourcegraph.com/search?q=context:global++/Linter%5C.Config%5C%7D/+count:all&patternType=keyword&case=yes&sm=0)

The type system allows for many great DevX improvements beyond just checking and enforcement.
If you use VSCode with JS -- you're relying on the type system provided by TypeScript without even knowing it!
- Member autocomplete? TypeScript
- Property autocomplete? TypeScript
- Imported name autocomplete? TypeScript
- JSDoc? TypeScript

All of these features are powered by TS's ability to model the code.

> Worse case there's an error if you access a string that doesn't exist

If you've got the opportunity to provide a better, more immediate feedback loop to users -- why would you not want to support that?
- "Write a magic string, run eslint, wait for eslint to error" is a slow workflow that's pretty painful.
- "Change a plugin reference, run eslint, wait for eslint to error" is painful.
- "Update a plugin version, run eslint, wait for eslint to error" is painful.

Providing these things in the type system means people can get immediate feedback in their IDE.
It means that plugin owners can provide JSDoc annotations to people to provide docs and links directly in their IDE.

--

By continuing on this path you are actively creating a worse DevX for users:
- it cannot be modeled with types -- meaning:
  - it is not auto-complete-able
  - it is not checkable pre-lint run
  - it cannot have JSDoc annotations added
- it is yet another way to do things:
  - shareable configs cannot use this system
  - not all plugins declare their configs on `.configs` -- these plugins won't work with this system
  - not all plugins declare _objects_ and instead declare functions -- these plugins won't work with this system

Reply:
The big thing the ecosystem has been asking for and looking for from you guys is clear documentation saying how to do things with flat configs. This is why there's a lack of standardisation -- because there wasn't clear documentation on how you thought people should do things.

And so in lieu of documentation the ecosystem made it up. And as expected - without direction people went in a lot of different directions.

If you want to standardise things -- clearly document the standard!
Clearly document the conventions in unambiguous terms.
Clearly document what properties are required and what is **_optional but recommended_**.
Clearly document that you discourage `/configs` import paths.
Heck - provide an example repo showing how you would recommend people structure things.

Reply:
I can just imagine how painful this is going to be for both users and plugin authors.

For example -- if a user does
```ts
import ts from 'typescript-eslint';

export default [
  { plugins: { ts: ts.plugin } },
  'ts/recommended',
];
```

They will get an error.
We do not export `ts.plugin.configs.recommended`.
We export `ts.configs.recommended`.

A user will be confused about why this didn't work.
They won't understand the differentiation.
They will file issues with us and we'll have to tell them to use `ts.configs.recommended` -- wasting everyone's time.

The same will happen for plugins that currently export a `.flatConfigs` property.
The same will happen for plugins that export config functions on `configs`.
The same will happen for plugins that have a `/configs` import path.

Reply:
@nzakas Please listen to @bradzacher on this. He represents our collective experience of working with these constraints for nearly a decade at this point.

It's worth reiterating - meaningfully more than half of your userbase uses eslint via typescript-eslint. There is also an above average representation of large scale, real world repos in that proportion, because those teams overwhelming use TypeScript. Any data from our userbase has has a disproportionately large amount of value to you.

The typescript-eslint team in our day jobs spends a significant amount of time shipping code in these exact kinds of repos, so as well as our own experiences, we also watch developers (both colleagues and S&P 500 customers, from the most junior to the most senior) get tripped up and frustrated by the exact things Brad is feeding back on.

This RFC is itself an admission that Brad's feedback should not have been dismissed the first time around when releasing flat config, please, heed the feedback he is giving you now

Reply:
First, I appreciate @bradzacher's passion for doing the right thing. While I don't believe all of his criticism is fair, I assure you that I and the team in general always read everything he posts. 

That said, I do want to remind you that we are getting feedback that the current way of doing things, allowing plugins to export configs however they want, is **not** working for our users. That is data that we have and feedback we've received repeatedly. Simply adding `extends` and flattening the array doesn't do anything to address that problem. It still requires users to find all the different ways that plugins are providing configs.

I also want to remind you that we, too, have experience working with teams inside of companies that are having problems. That's why I put this RFC together. The feedback you're receiving and providing back to us is valuable, and so is the feedback we are receiving directly from our users and sponsors. We need to be balanced in assessing that feedback across the board.

> This RFC is itself an admission that Brad's feedback should not have been dismissed the first time around when releasing flat config, please, heed the feedback he is giving you now

There was never any dismissal of Brad's feedback. His feedback came very late in the development cycle and so it wasn't prudent to start making changes before we got more feedback from the community at large. Now that we have, we're looking to make changes. I think that's an indication that the process is working.